File: app/src/main/java/tpcreative/co/qrscanner/ui/create/BarcodeFragment.java
package tpcreative.co.qrscanner.ui.create;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import androidx.appcompat.widget.Toolbar;
import com.basgeekball.awesomevalidation.AwesomeValidation;
import com.basgeekball.awesomevalidation.ValidationStyle;
import com.basgeekball.awesomevalidation.utility.RegexTemplate;
import com.basgeekball.awesomevalidation.utility.custom.SimpleCustomValidation;
import com.google.zxing.BarcodeFormat;
import com.google.zxing.client.result.ParsedResultType;
import org.apache.commons.validator.routines.checkdigit.EAN13CheckDigit;
import butterknife.BindView;
import butterknife.OnClick;
import tpcreative.co.qrscanner.R;
import tpcreative.co.qrscanner.common.Navigator;
import tpcreative.co.qrscanner.common.SingletonGenerate;
import tpcreative.co.qrscanner.common.SingletonSave;
import tpcreative.co.qrscanner.common.Utils;
import tpcreative.co.qrscanner.common.activity.BaseActivitySlide;
import tpcreative.co.qrscanner.model.Create;
import tpcreative.co.qrscanner.model.EnumImplement;
import tpcreative.co.qrscanner.model.FormatTypeModel;
import tpcreative.co.qrscanner.common.entities.SaveEntity;
import tpcreative.co.qrscanner.model.SaveModel;

public class BarcodeFragment extends BaseActivitySlide implements SingletonGenerate.SingletonGenerateListener,GenerateView {

    private static final String TAG = BarcodeFragment.class.getSimpleName();
    AwesomeValidation mAwesomeValidation ;
    @BindView(R.id.edtText)
    EditText editText;
    @BindView(R.id.spinner)
    Spinner spinner;
    private SaveModel save;
    private GeneratePresenter presenter;
    private ArrayAdapter<FormatTypeModel> dataAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_barcode);
        final Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        presenter = new GeneratePresenter();
        presenter.bindView(this);
        presenter.doInitView();
        Bundle bundle = getIntent().getExtras();
        final SaveModel mData = (SaveModel) bundle.get(getString(R.string.key_data));
        if (mData!=null){
            save = mData;
            onSetData();
        }
        else{
            Utils.Log(TAG,"Data is null");
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_select, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.menu_item_select:{
                if (mAwesomeValidation.validate()){
                    Utils.Log(TAG,"Passed");
                    Create create = new Create();
                    create.productId = editText.getText().toString().trim();
                    create.createType = ParsedResultType.PRODUCT;
                    create.barcodeFormat = presenter.mType.name();
                    create.enumImplement = (save != null) ? EnumImplement.EDIT : EnumImplement.CREATE ;
                    create.id = (save != null) ? save.id : 0 ;
                    Navigator.onMoveToReview(this,create);
                }
                else{
                    Utils.Log(TAG,"error");
                }
                return true;
            }
        }
        return super.onOptionsItemSelected(item);
    }

    @OnClick(R.id.btnRandom)
    public void onClickedRandom(View view){
        String mValue = Utils.generateRandomDigits(presenter.mLength-1) + "";
        String mResult = Utils.generateEAN(mValue);
        editText.setText(mResult);
    }

    private void addValidationForEditText() {
        mAwesomeValidation.addValidation(this,R.id.edtText, RegexTemplate.NOT_EMPTY,R.string.err_text);
        mAwesomeValidation.addValidation(this, R.id.edtText, new SimpleCustomValidation() {
            @Override
            public boolean compare(String input) {
                // check if the age is >= 18
                if (presenter.mType == BarcodeFormat.EAN_13){
                    if (input.length()==13){
                        if (EAN13CheckDigit.EAN13_CHECK_DIGIT.isValid(input)){
                            return true;
                        }
                        return false;
                    }else{
                        return  false;
                    }
                }
               return true;
            }
        }, R.string.warning_barcode_length_13);
        mAwesomeValidation.addValidation(this, R.id.edtText, new SimpleCustomValidation() {
            @Override
            public boolean compare(String input) {
                // check if the age is >= 18
                String mValue = Utils.checkSum(input) + "";
                Utils.Log(TAG,mValue);
                if (presenter.mType == BarcodeFormat.EAN_8){
                    if (input.length()==8){
                        if (Utils.checkGTIN(input)){
                            return true;
                        }
                        return false;
                    }else{
                        return  false;
                    }
                }
                return true;
            }
        }, R.string.warning_barcode_length_8);
    }

    public void FocusUI(){
        editText.requestFocus();
    }

    public void onSetData(){
        editText.setText(save.text);
        if (save.createType.equals(ParsedResultType.PRODUCT.name())){
            if (save.barcodeFormat.equals(BarcodeFormat.EAN_13.name())){
                presenter.mType = BarcodeFormat.EAN_13;
                presenter.mLength = 13;
                spinner.setSelection(0);
            }
            else if (save.barcodeFormat.equals(BarcodeFormat.EAN_8.name())){
                presenter.mType = BarcodeFormat.EAN_8;
                presenter.mLength = 8;
                spinner.setSelection(1);
            }
        }
    }


    @Override
    public void onStart() {
        super.onStart();
        Utils.Log(TAG,"onStart");
        mAwesomeValidation =  new AwesomeValidation(ValidationStyle.BASIC);
        addValidationForEditText();
        if (save!=null){
            onSetData();
        }
        FocusUI();
    }

    @Override
    public void onStop() {
        super.onStop();
        Utils.Log(TAG,"onStop");
    }

    @Override
    public void onPause() {
        super.onPause();
        Utils.Log(TAG,"onPause");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        SingletonGenerate.getInstance().setListener(null);
        Log.d(TAG,"onDestroy");
    }

    @Override
    public void onResume() {
        super.onResume();
        SingletonGenerate.getInstance().setListener(this);
        Log.d(TAG,"onResume");
    }

    @Override
    public void onCompletedGenerate() {
        SingletonSave.getInstance().reLoadData();
        Utils.Log(TAG,"Finish...........");
        finish();
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == Activity.RESULT_OK && requestCode == Navigator.CREATE) {
            Utils.Log(TAG,"Finish...........");
            SingletonSave.getInstance().reLoadData();
            finish();
        }
    }

    @Override
    public Context getContext() {
        return this;
    }

    @Override
    public void onSetView() {
        dataAdapter.notifyDataSetChanged();
    }

    @Override
    public void onInitView() {
        addItemsOnSpinner();
        addListenerOnSpinnerItemSelection();
        presenter.getBarcodeFormat();
    }

    // add items into spinner dynamically
    public void addItemsOnSpinner() {
        dataAdapter = new ArrayAdapter<FormatTypeModel>(this,
                android.R.layout.simple_spinner_item, presenter.mBarcodeFormat);
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(dataAdapter);
    }

    public void addListenerOnSpinnerItemSelection() {
        spinner.setOnItemSelectedListener(new CustomOnItemSelectedListener());
    }

    public class CustomOnItemSelectedListener implements AdapterView.OnItemSelectedListener {
        public void onItemSelected(AdapterView<?> parent, View view, int pos,long id) {
            FormatTypeModel type = dataAdapter.getItem(pos);
            if (type.id == BarcodeFormat.EAN_13.name()){
                editText.setHint(R.string.hint_13);
                presenter.doSetMaxLength(true,editText);
            }else{
                editText.setHint(R.string.hint_8);
                presenter.doSetMaxLength(false,editText);
            }
            presenter.mType = BarcodeFormat.valueOf(type.id);
        }
        @Override
        public void onNothingSelected(AdapterView<?> arg0) {
            // TODO Auto-generated method stub
        }
    }
}

File: app/src/main/java/tpcreative/co/qrscanner/ui/create/BarcodeFragment.kt
package tpcreative.co.qrscanner.ui.create
import android.content.*
import android.os.Bundle
import android.view.*
import android.widget.AdapterView
import android.widget.ArrayAdapter
import androidx.appcompat.widget.*
import com.basgeekball.awesomevalidation.AwesomeValidation
import com.basgeekball.awesomevalidation.ValidationStyle
import com.basgeekball.awesomevalidation.utility.RegexTemplate
import com.basgeekball.awesomevalidation.utility.custom.SimpleCustomValidation
import com.google.gson.Gson
import com.google.zxing.BarcodeFormat
import com.google.zxing.client.result.ParsedResultType
import kotlinx.android.synthetic.main.fragment_barcode.*
import org.apache.commons.validator.routines.checkdigit.EAN13CheckDigit
import tpcreative.co.qrscanner.R
import tpcreative.co.qrscanner.common.*
import tpcreative.co.qrscanner.common.activity.BaseActivitySlide
import tpcreative.co.qrscanner.model.*
import tpcreative.co.qrscanner.viewmodel.GenerateViewModel

class BarcodeFragment : BaseActivitySlide(), GenerateSingleton.SingletonGenerateListener {
    var mAwesomeValidation: AwesomeValidation? = null
    var save: SaveModel? = null
    var dataAdapter: ArrayAdapter<FormatTypeModel>? = null
    lateinit var viewModel : GenerateViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_barcode)
        initUI()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_select, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_item_select -> {
                if (mAwesomeValidation?.validate() == true) {
                    Utils.Log(TAG, "Passed")
                    val create = Create(save)
                    create.productId = edtText.text.toString().trim { it <= ' ' }
                    create.barcodeFormat = viewModel.mType?.name
                    if (create.barcodeFormat == BarcodeFormat.EAN_8.name || create.barcodeFormat == BarcodeFormat.EAN_13.name || create.barcodeFormat == BarcodeFormat.UPC_A.name || create.barcodeFormat == BarcodeFormat.UPC_E.name){
                        create.createType = ParsedResultType.PRODUCT
                    }else{
                        create.createType = ParsedResultType.TEXT
                    }
                    Navigator.onMoveToReview(this, create)
                } else {
                    Utils.Log(TAG, "error")
                }
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun addValidationForEditText() {
        mAwesomeValidation?.addValidation(this, R.id.edtText, RegexTemplate.NOT_EMPTY, R.string.err_text)
        mAwesomeValidation?.addValidation(this, R.id.edtText, SimpleCustomValidation { input -> // check if the age is >= 18
            if (viewModel.mType == BarcodeFormat.EAN_13) {
                if (input.length == 13) {
                    return@SimpleCustomValidation EAN13CheckDigit.EAN13_CHECK_DIGIT.isValid(input)
                } else {
                    return@SimpleCustomValidation false
                }
            }
            true
        }, R.string.warning_barcode_length_13)
        mAwesomeValidation?.addValidation(this, R.id.edtText, SimpleCustomValidation { input -> // check if the age is >= 18
            if (viewModel.mType == BarcodeFormat.EAN_8) {
                if (input.length == 8) {
                    return@SimpleCustomValidation Utils.checkGTIN(input)
                } else {
                    return@SimpleCustomValidation false
                }
            }
            true
        }, R.string.warning_barcode_length_8)

        mAwesomeValidation?.addValidation(this, R.id.edtText, SimpleCustomValidation { input -> // check if the age is >= 18
            if (viewModel.mType == BarcodeFormat.UPC_E) {
                if (input.length == 8) {
                    return@SimpleCustomValidation Utils.checkGTIN(input)
                } else {
                    return@SimpleCustomValidation false
                }
            }
            true
        }, R.string.warning_barcode_UPC_E_length_8)
        mAwesomeValidation?.addValidation(this, R.id.edtText, SimpleCustomValidation { input -> // check if the age is >= 18
            if (viewModel.mType == BarcodeFormat.UPC_A) {
                if (input.length == 12) {
                    return@SimpleCustomValidation Utils.checkGTIN(input)
                } else {
                    return@SimpleCustomValidation false
                }
            }
            true
        }, R.string.warning_barcode_UPC_A_length_12)
        mAwesomeValidation?.addValidation(this, R.id.edtText, SimpleCustomValidation { input -> // check if the age is >= 18
            if (viewModel.mType == BarcodeFormat.ITF) {
                return@SimpleCustomValidation Utils.checkITF(input)
            }
            true
        }, R.string.warning_barcode_ITF)

    }

    private fun focusUI() {
        edtText.requestFocus()
    }

    fun onSetData() {
        edtText.setText(save?.text)
        if (save?.createType == ParsedResultType.PRODUCT.name) {
            if (save?.barcodeFormat == BarcodeFormat.EAN_13.name) {
                viewModel.mType = BarcodeFormat.EAN_13
                viewModel.mLength = 13
                spinner.setSelection(0)
            } else if (save?.barcodeFormat == BarcodeFormat.EAN_8.name) {
                viewModel.mType = BarcodeFormat.EAN_8
                viewModel.mLength = 8
                spinner.setSelection(1)
            }else if (save?.barcodeFormat == BarcodeFormat.UPC_E.name){
                viewModel.mType = BarcodeFormat.UPC_E
                viewModel.mLength = 8
                spinner.setSelection(1)
            }else if (save?.barcodeFormat == BarcodeFormat.UPC_A.name){
                viewModel.mType = BarcodeFormat.UPC_A
                viewModel.mLength = 12
                spinner.setSelection(1)
            }else if (save?.barcodeFormat == BarcodeFormat.CODABAR.name){
                viewModel.mType = BarcodeFormat.CODABAR
                viewModel.mLength = 40
                spinner.setSelection(1)
            }else if (save?.barcodeFormat == BarcodeFormat.DATA_MATRIX.name){
                viewModel.mType = BarcodeFormat.DATA_MATRIX
                viewModel.mLength = 50
                spinner.setSelection(1)
            }
            else if (save?.barcodeFormat == BarcodeFormat.PDF_417.name){
                viewModel.mType = BarcodeFormat.PDF_417
                viewModel.mLength = 50
                spinner.setSelection(1)
            }
            else if (save?.barcodeFormat == BarcodeFormat.AZTEC.name){
                viewModel.mType = BarcodeFormat.AZTEC
                viewModel.mLength = 50
                spinner.setSelection(1)
            }
            else if (save?.barcodeFormat == BarcodeFormat.CODE_128.name){
                viewModel.mType = BarcodeFormat.CODE_128
                viewModel.mLength = 50
                spinner.setSelection(1)
            }
            else if (save?.barcodeFormat == BarcodeFormat.CODE_39.name){
                viewModel.mType = BarcodeFormat.CODE_39
                viewModel.mLength = 50
                spinner.setSelection(1)
            }
            else if (save?.barcodeFormat == BarcodeFormat.CODE_93.name){
                viewModel.mType = BarcodeFormat.CODE_93
                viewModel.mLength = 50
                spinner.setSelection(1)
            }
            else if (save?.barcodeFormat == BarcodeFormat.ITF.name){
                viewModel.mType = BarcodeFormat.ITF
                viewModel.mLength = 50
                spinner.setSelection(1)
            }
            Utils.Log(TAG,"Save ${Gson().toJson(save)}")
        }
    }

    public override fun onStart() {
        super.onStart()
        Utils.Log(TAG, "onStart")
        mAwesomeValidation = AwesomeValidation(ValidationStyle.BASIC)
        addValidationForEditText()
        focusUI()
    }

    public override fun onStop() {
        super.onStop()
        Utils.Log(TAG, "onStop")
    }

    public override fun onPause() {
        super.onPause()
        Utils.Log(TAG, "onPause")
    }

    public override fun onDestroy() {
        super.onDestroy()
        GenerateSingleton.getInstance()?.setListener(null)
        Utils.Log(TAG, "onDestroy")
    }

    public override fun onResume() {
        super.onResume()
        GenerateSingleton.getInstance()?.setListener(this)
        Utils.Log(TAG, "onResume")
    }

    override fun onCompletedGenerate() {
        SaveSingleton.getInstance()?.reloadData()
        Utils.Log(TAG, "Finish...........")
        finish()
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == Navigator.CREATE) {
            Utils.Log(TAG, "Finish...........")
            SaveSingleton.getInstance()?.reloadData()
            finish()
        }
    }

    fun onSetView() {
        dataAdapter?.notifyDataSetChanged()
    }

    fun onInitView() {
        addItemsOnSpinner()
        addListenerOnSpinnerItemSelection()
        getBarcodeFormat()
    }

    // add items into spinner dynamically
    private fun addItemsOnSpinner() {
        dataAdapter = ArrayAdapter(this,
                android.R.layout.simple_spinner_item, viewModel.mBarcodeFormat)
        dataAdapter?.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = dataAdapter
    }

    private fun addListenerOnSpinnerItemSelection() {
        spinner.onItemSelectedListener = CustomOnItemSelectedListener()
    }

    inner class CustomOnItemSelectedListener : AdapterView.OnItemSelectedListener {
        override fun onItemSelected(parent: AdapterView<*>?, view: View?, pos: Int, id: Long) {
            val type = dataAdapter?.getItem(pos)
            if (type?.id === BarcodeFormat.EAN_13.name) {
                edtText.setHint(R.string.hint_13)
                viewModel.doSetMaxLength(BarcodeFormat.EAN_13, edtText)
            }
            else if (type?.id === BarcodeFormat.EAN_8.name){
                edtText.setHint(R.string.hint_8)
                viewModel.doSetMaxLength(BarcodeFormat.EAN_8, edtText)
            }
            else if (type?.id === BarcodeFormat.UPC_E.name){
                edtText.setHint(R.string.hint_8)
                viewModel.doSetMaxLength(BarcodeFormat.UPC_E, edtText)
            }else if (type?.id === BarcodeFormat.UPC_A.name){
                edtText.setHint(R.string.hint_12)
                viewModel.doSetMaxLength(BarcodeFormat.UPC_A, edtText)
            }
            else if (type?.id === BarcodeFormat.CODABAR.name){
                edtText.setHint(R.string.hint_digits)
                viewModel.doSetMaxLength(BarcodeFormat.CODABAR, edtText)
            }
            else if (type?.id === BarcodeFormat.CODE_39.name){
                edtText.setHint(R.string.hint_uppercase_characters)
                viewModel.doSetMaxLength(BarcodeFormat.CODE_39, edtText)
            }
            else if (type?.id === BarcodeFormat.CODE_93.name){
                edtText.setHint(R.string.hint_uppercase_characters)
                viewModel.doSetMaxLength(BarcodeFormat.CODE_93, edtText)
            }
            else if (type?.id === BarcodeFormat.ITF.name){
                edtText.setHint(R.string.hint_digits)
                viewModel.doSetMaxLength(BarcodeFormat.ITF, edtText)
            }
            else {
                edtText.setHint(R.string.hint_characters)
                viewModel.doSetMaxLength(BarcodeFormat.DATA_MATRIX, edtText)
            }
            viewModel.mType = BarcodeFormat.valueOf(type?.id ?:"")
            if (viewModel.isText(save?.text)){
                edtText.setText(save?.text)
                edtText.requestFocus()
            }
        }

        override fun onNothingSelected(arg0: AdapterView<*>?) {
            // TODO Auto-generated method stub
        }
    }

    companion object {
        private val TAG = BarcodeFragment::class.java.simpleName
    }
}